source/common/stream_info/utility.cc

```diff
@@ -134,5 +140,215 @@ Utility::formatDownstreamAddressJustPort(const Network::Address::Instance& addre
   return port;
 }
 
+const absl::optional<Http::Code>
+ProxyStatusUtils::recommendedHttpStatusCode(const ProxyStatusError proxy_status) {
+  // This switch statement was derived from the mapping from proxy error type to
+  // recommended HTTP status code in
+  // https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-proxy-status-05#section-2.3 and below.
+  //
+  // TODO(ambuc): Replace this with the non-draft URL when finalized.
+  switch (proxy_status) {
+  case ProxyStatusError::DnsTimeout:
+  case ProxyStatusError::ConnectionTimeout:
+  case ProxyStatusError::ConnectionReadTimeout:
+  case ProxyStatusError::ConnectionWriteTimeout:
+  case ProxyStatusError::HttpResponseTimeout:
+    return Http::Code::GatewayTimeout; // 504
+  case ProxyStatusError::DnsError:
+  case ProxyStatusError::DestinationIpProhibited:
+  case ProxyStatusError::DestinationIpUnroutable:
+  case ProxyStatusError::ConnectionRefused:
+  case ProxyStatusError::ConnectionTerminated:
+  case ProxyStatusError::TlsProtocolError:
+  case ProxyStatusError::TlsCertificateError:
+  case ProxyStatusError::TlsAlertReceived:
+  case ProxyStatusError::HttpResponseIncomplete:
+  case ProxyStatusError::HttpResponseHeaderSectionSize:
+  case ProxyStatusError::HttpResponseHeaderSize:
+  case ProxyStatusError::HttpResponseBodySize:
+  case ProxyStatusError::HttpResponseTrailerSectionSize:
+  case ProxyStatusError::HttpResponseTrailerSize:
+  case ProxyStatusError::HttpResponseTransferCoding:
+  case ProxyStatusError::HttpResponseContentCoding:
+  case ProxyStatusError::HttpUpgradeFailed:
+  case ProxyStatusError::HttpProtocolError:
+  case ProxyStatusError::ProxyLoopDetected:
+    return Http::Code::BadGateway; // 502
+  case ProxyStatusError::DestinationNotFound:
+  case ProxyStatusError::ProxyInternalError:
+  case ProxyStatusError::ProxyConfigurationError:
+    return Http::Code::InternalServerError; // 500
+  case ProxyStatusError::DestinationUnavailable:
+  case ProxyStatusError::ConnectionLimitReached:
+    return Http::Code::ServiceUnavailable; // 503
+  case ProxyStatusError::HttpRequestDenied:
+    return Http::Code::Forbidden; // 403
+  case ProxyStatusError::ProxyInternalResponse:
+  case ProxyStatusError::HttpRequestError:
+  default:
+    return absl::nullopt;
+  }
+}
+
+const std::string
+ProxyStatusUtils::toString(const StreamInfo& stream_info, const ProxyStatusError error,
```

Comment:
nit: maybe make this method name more descriptive as it is not clear what exactly is converted to string.

Reply:
Done.